delegated_credentials.go

```diff
@@ -0,0 +1,295 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"time"
+)
+
+// After only renewThreshold of credential's validTime is left a new one is created
+const renewThreshold = 0.2
+
+// The Delegation Usage x509 extension identifier - temporary, TBD
+var delegatedCredentialsIdentifier = asn1.ObjectIdentifier{2, 5, 29, 99}
+
+type GetCertificate func(*ClientHelloInfo) (*Certificate, error)
+
+type DelegatedCredential struct {
+	ValidTime int64
+	PublicKey interface{}
+}
+
+type cachedCredential struct {
+	credential []byte
+	validTime  int64
+	privateKey crypto.PrivateKey
+	renewAfter time.Time
+}
+
+type CredentialsConfig struct {
+	Cert      *Certificate
+	Validity  time.Duration
+	scheme    SignatureScheme
+	version   uint16
+	timeFunc  func() time.Time
+	renewChan chan struct{}
+}
+
+var cachedCredentials map[SignatureScheme]*cachedCredential
+
+// Creates new GetCertificate function which modifies the certificate
+// to include a delegated credential.
+func NewDelegatedCredentialsGetCertificate(config *CredentialsConfig) GetCertificate {
+	return func(clientHelloInfo *ClientHelloInfo) (*Certificate, error) {
+		if !isCertificateValidForDelegationUsage(config.Cert.Leaf) {
+			return nil, errors.New("tls: Delegated Credentials not supported by the certificate (DelegationUsage extension missing)")
+		}
+		selectedScheme := selectSignatureScheme(clientHelloInfo.SignatureSchemes, config.Cert.Leaf)
+		if selectedScheme == 0 {
+			return nil, errors.New("tls: No valid signature scheme found for Delegated Credentials")
+		}
+		version := selectVersion(clientHelloInfo.SupportedVersions)
+		if version == 0 {
+			return nil, errors.New("tls: Only TLS 1.2 or 1.3 are supported")
+		}
+		if config.timeFunc == nil {
+			config.timeFunc = time.Now
+		}
+
+		config.scheme = selectedScheme
+		config.version = version
+		err := fetchCredentialFromCache(config)
+		if err != nil {
+			return nil, err
+		}
+
+		config.Cert.DelegatedCredential = cachedCredentials[selectedScheme].credential
+		config.Cert.PrivateKey = cachedCredentials[selectedScheme].privateKey
+		return config.Cert, nil
+	}
+}
+
+// Checks if a credential is already stored in memory. If not or the credential
+// is already expired a new credential is created.
+func fetchCredentialFromCache(config *CredentialsConfig) (err error) {
+	if cachedCredentials == nil {
+		cachedCredentials = make(map[SignatureScheme]*cachedCredential)
+	}
+	if cachedCredentials[config.scheme] == nil {
+		cachedCredentials[config.scheme], err = newCredential(config)
+	} else {
+		if config.timeFunc().After(cachedCredentials[config.scheme].renewAfter) {
+			go func() {
+				cachedCredentials[config.scheme], err = newCredential(config)
+				if config.renewChan != nil {
+					config.renewChan <- struct{}{}
```

Comment:
Only in [tests](https://github.com/cloudflare/tls-tris/pull/29/files#diff-74969c972e0630d3c7b131f8f82325e9R270). I was reluctant to add the channel just because of tests, but I wanted to test that the new credential is created, which I can do only after the goroutine finishes. I was talking to Salmen about this and he advised me this.